### DISCUSSION TITLE
Single components rendered as SSG in SSR pages

### DISCUSSION THREAD

**Discussion by ziriuz84** [withastro/roadmap#563] [Proposal]
### Body

# Summary

It will be useful to have possibility to render as SSG single components in SSR pages and viceversa

# Background & Motivation

Until today we can choose between SSR and SSG only in Astro configuration, and it's ok. If I use SSR I can decide to render as SSG only the entire page, but there may be shared components between SSR and SSG  pages or components that doesn't really need to be SSR, like headers, footers or navigation. It will be very useful to have the choice on how to render the single component, with benefits in velocity and modularity

# Goals

- Possibility to choose between SSG and SSR for the single component
- Better modularity in rendering components
- Sometimes there are single components where SSR brings no benefits.

# Example

A solution could be to add a directive in pros, like 
`<Component render:ssg />`

Another solution could be to use the same method used for pages in frontmatter, something like
`export const prerender = true;`

**Comment by LopezNoah** [comment]
This reminds of another proposal: [Server-side islands with server:dynamic](https://github.com/withastro/roadmap/discussions/116).
Maybe also tied into this: [Container API: render components in isolation](https://github.com/withastro/roadmap/discussions/462).
Astro can already render out components with SSG, but if you could fetch components from the server on-demand then it would open this up. Some directives could look like: ```server:dynamic```, ```server:static```, ```render:ssg```, or ```render:ssr```.

Since Astro can already return Responses, the client can make a fetch call for the dynamic data on an SSG page. The client just needs to know where to place those components/responses. This could be done with empty divs that get replaced as shown below:

## foo.astro
```astro
---
import AstroHeart from "../components/AstroHeart.astro";
import Layout from "../layouts/Layout.astro";

export const prerender = true;
---

<Layout title="Welcome to Astro.">
	<main>
        <AstroHeart />
        <div id="AstroHeart">Loading...</div>
    </main>
    <script is:inline>
        const element = document.getElementById('AstroHeart');

        if (element != null){
            fetch("/api/AstroHeart")
            .then(response => response.text())
            .then(data => {
                setTimeout(() => {
                    element.innerHTML = data;
                }, 1500);
            })
            .catch(error => {
                console.error('Error:', error);
            });
        }
    </script>
</Layout>
```

## api/[component].astro
```astro
---
import AstroHeart from "../../components/AstroHeart.astro";

export const prerender = false;

const { component } = Astro.params;

if (!component) {
    throw new Error();
}

console.log(component);

Astro.response.headers.set("Content-Type", "text/html; charset=utf-8");
---

{component == 'AstroHeart' ?
 (<AstroHeart />) : null
}
<div>Hey there</div>
```

This isn't the best experience imo. Maybe the div above should really be it's own Astro component that receives props, but it renders out to just the div. The "foo" route is static, and then the server just swaps in the other piece. I had some problems using AstroInstance and the default property, so I just explicitly declared the component in the file. [AstroHeart](https://docs.astro.build/en/guides/client-side-scripts/#web-components-with-custom-elements) comes from the docs btw

**Reply by LopezNoah** [reply]
It'd be cool to have the route compile the dynamic components into one big Promise so we avoid waterfalls. Astro now has parallel rendering as of v2.5 so maybe that could handle it for the user already!

**Reply by ziriuz84** [reply]
This code tastes a little like a "workaround", but it's useful for now. But I also think that it's not the best experience for the devoloper. A prop to set the render mode will be more easy and useful also for the opposite, i.e. render statically a component inside a ssr page.

**Reply by LopezNoah** [reply]
Completely agree. It's definitely a workaround and doesn't work the other way around (Render static component in a server rendered page). Here's a more polished example I whipped up: [Link](https://github.com/LopezNoah/Astro-Dynamic/blob/029635df09ebf7a81f7e4a49b635ee3fde9d49b7/src/components/Dynamic.astro). Same concept, but now the user can just use ```<Dynamic />``` with some props in their Astro file.

This is still way too much for someone to hand roll, and would be awesome for the server to know to do these things at run-time. Without injecting the hydration script into the head, the interactive components don't load. The code is a bit funky, and it didn't even work for me in build (only for dev) but you can see the idea.

For static components, would those just be generated as their own HTML pages? That way the server doesn't generate them, but the components have their own special route at build time.
